tests/packages/mlx/test_mellanox-modules-compat.py

```diff
@@ -0,0 +1,27 @@
+# Requirements:
+# From --hosts parameter:
+# - host(A1): any master host of a pool, with access to XCP-ng RPM repositories.
+import logging
+
+from pkgfixtures import host_with_saved_yum_state
+
+MLX4_MODULE = 'mlx4_en'
+
+def load_unload_mlx_module(host):
+    logging.info("loading mlx module")
+    host.ssh(['modprobe -v', MLX4_MODULE])
+    logging.info("unloading mlx module")
+    host.ssh(['modprobe -r', MLX4_MODULE])
+
+def test_install_mlx_modules_alt(host_with_saved_yum_state):
+    host = host_with_saved_yum_state
+    # Ensure the modules are unloaded
+    host.yum_remove(['mlx4-modules-alt', 'mellanox-mlnxen-alt'])
+
+    # Start by loading mlx4
+    host.yum_install(['mlx4-modules-alt'])
+    load_unload_mlx_module(host)
```

Comment:
This doesn't unload `mlx_compat.ko` to make space for the one provided by `mellanox-mlnxen-alt` in the next part of the text.

Reply:
modprobe -r remove also dependencies no? So it should remove `mlx4_en` and its dependencies...
Or are you saying that `mlx_compat.ko` is already installed and so not installed as a dependency? So it should be removed at the beginning of the test. I looked on my host and it is not loaded after running test. But I have not a clean host so I will test again using fresh install.

Reply:
I think you're right. But, yes, `mlx_compat.ko` is already installed in the case of XCP-ng 8.3. Not loaded if there's no hardware that requires it, but installed it is.

Reply:
Just tested on my system and I have:
```
[18:05 xcp-gtn-cirepo ~]# modprobe -v mlx4_en
insmod /lib/modules/4.19.0+1/kernel/net/core/devlink.ko
insmod /lib/modules/4.19.0+1/updates/mlx_compat.ko
insmod /lib/modules/4.19.0+1/override/mlx4_core.ko
insmod /lib/modules/4.19.0+1/override/mlx4_en.ko
```
```
[18:06 xcp-gtn-cirepo ~]# modprobe -rv mlx4_en
rmmod mlx4_en
rmmod mlx4_core
rmmod devlink
rmmod mlx_compat
```

Reply:
Note that I have chosen at fist to load `mlx4_ib` because it loads all dependencies including `mlx4_en`.

Reply:
And on my fresh install of xcp-ng 8.3 `mlx_compat`  was not loaded. But ok I can check and ensure that it is not loaded before testing. But does that mean that if it is already loaded before testing I need to load it when test ends. And the other way, if it was not loaded it should remains unloaded? or is it too much?

Reply:
On hardware which uses the driver, we should not unload it, so I suggest you just add a fixture which checks that the module is not loaded and fails the test otherwise.

Reply:
Or, since there's just one test, assert at the beginning of the test, without taking any measures to solve it directly. If this causes issues in the future, we can think of something more elaborate.

Reply:
Let's also assert between the two parts of the test.

Reply:
> On hardware which uses the driver, we should not unload it, so I suggest you just add a fixture which checks that the module is not loaded and fails the test otherwise.

but it is only use for testing so we know on which hardware we are running right? I will check if the module is not loaded and if it is we will fail. We will update the test in the future if needed as you suggest.

Reply:
We know our current hardware, not future hardware, nor on what hardware someone could want to run the tests.

Reply:
I finally unload the mlx_compat module and reload it at the end if needed. But if you prefer to just failed if the module is loaded I can just assert that the module is not loaded. I tested both condition (module already loaded before the test and not loaded before the test). It looks ok.

Reply:
Will this work if `mlx5_core` is loaded and depends on it to operate the network?

Reply:
If `mlx5_core` is loaded the test will failed while trying to unload `mlx_compat`. So the test will fail and the network will continue to work. The command `modprobe -r mlx_compat` cannot succeed if the module is in use by another module (like mlx5_core).